gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+propose an alternative chain filtering mechanism for light clients. Our
+proposal sports a greater degree of privacy than BIP 37 as filtering is now
+done on the client side. Clients download a deterministically generated filter
+for a block and query it locally. If relevant items are found in the filter
+then the ''entire'' block will be fetched. The decoupling of filter querying from
+active communication with full-nodes enables light clients to fetch blocks from
+''any'' source. Extremely privacy conscious light clients may opt to anonymously
+fetch blocks using cryptographic techniques such a Private Information
+Retrieval [5].
+
+In order to reduce the size of the filter, we use a data structure capable of
+probabilistic set membership. We elide the selection of the bloom filter data
+structure in favor of utilising Golomb-Rice coding which allows us to generate
+filters more compact than bloom filters which approach the theoretical minimum
+size for probabilistic data structure.
+
+Light clients operating using the method described in this document are able to
+verify the authenticity of filters received, thereby eliminating the ability
+for full-nodes to lie by omission. Such client side filtering also improves the
+utility of light clients for generic applications beyond simple wallets and
+usage of basic public key templates. 
+
+Finally, full-nodes only need to construct filters ''once'' as they're
+deterministically generated for each block. Once the index is built, no further
+active processing is required to serve light clients. Servicing light clients
+simply entails reading pre-computed filters and headers from disk and sending
+them over the network.
+
+
+== Preliminaries ==
+
+Before we specify the details of our proposal, we'll first go over a few
+preliminaries which will aid in the understanding our proposal.
+
+By <code>[]byte</code> we refer to a slice (or array) of bytes. This value is
+typically expressed in C-like languages as an array of uint_8's.
+
+By <code>Var-Int</code> we refer to the variable length integer encoding used
+widely within the Bitcoin p2p protocol as an efficient way to encode the number
+of items in a repeated series of items. The p2p message extensions in this
+proposal will utilize this variable-length integer encoding in an identical
+manner to the existing Bitcoin p2p messages.
+
+By <code>siphash(k, n)</code> we refer to an invocation of the <code>SipHash</code>
+pseudo-random function with <code>k</code> as the 128-bit key, and <code>n</code>
+as the input to the PRF. We instantiate with the recommended parameters of
+<code>c = 2</code> and <code>d = 4</code>.
+
+We define the concept of an abstract bit stream instantiated by the function:
+<code>new_bit_stream</code> The <code>bit_stream</code> has two functions that
+operate on it, <code>unary_encode(stream, n)</code> and
+<code>write_bits_big_endian(stream, n, k)</code> where <code>unary_encode(steam,
```

Comment:
I think we need to much more seriously consider whether we want to, for the first time, introduce big endian at the purely-p2p layer - it is nice at a bit level, but it kinda sucks to have yet *more* endianness confusion in Bitcoin's P2P protocol.

Reply:
Big endian just seemed natural initially. It's not as if we're using big-endian to encode a field that was _previously_ encoded using little-endian. This is a new data-structure all together.